std/numeric.d

```diff
@@ -3329,6 +3329,118 @@ C swapRealImag(C)(C input)
     return C(input.im, input.re);
 }
 
+/**
+Finds the mean (colloquially known as the average) of a range. If `r` does
+not provide the `length` member, then this function will do element by
+element summation, rather than the more accurate methods provided by
+$(REF sum, std,algorithm,iteration).
+
+An optional parameter `seed` is used for non-builtin types in order to
+correctly seed the summation with the equivalent to `0`.
+
+This function will return `real.nan` if the range is empty and if the element
+type of `r` is a built-in numerical type. However, if the seed value is defined,
+then empty ranges will return `ElementType!R.init`.
+
+This function is $(BIGOH r.length).
+
+Params:
+    r = An $(REF_ALTTEXT input range, isInputRange, std,range,primitives)
+    seed = An optional parameter for taking the mean of types which can't
+    implicitly convert to `real`. This value will initialize the summation,
+    so it should be equivalent to `0`.
+
+Returns:
+    the mean of r
+*/
+real mean(R)(R r)
+    if (isInputRange!R &&
+        is(ElementType!R : real) &&
+        !isInfinite!R)
+{
+    if (r.empty)
+    {
+        return real.nan;
+    }
+
+    return mean(r, real(0.0));
+}
+
+/// ditto
+auto mean(R, T)(R r, T seed)
+    if (isInputRange!R &&
+        (is(T == real) || !isNumeric!(T)) &&
+        is(typeof(seed + r.front)) &&
+        is(typeof(r.front / size_t(1))) &&
+        !isInfinite!R)
+{
+    import std.algorithm.iteration : sum, reduce;
+
+    if (r.empty)
+    {
+        return Unqual!(ElementType!R).init;
+    }
+
+    static if (hasLength!R)
+    {
+        if (r.length == 1)
+        {
+            return r.front;
+        }
+
+        return sum(r, seed) / r.length;
+    }
+    else
+    {
+        import std.typecons : tuple;
+
+        auto pair = reduce!((a, b) => tuple(a[0] + 1, a[1] + b))
+            (tuple(size_t(0), seed), r);
+
+        return pair[1] / pair[0];
```

Comment:
Please let's use a more efficient algorithm here. Or to quote [Jonis alonen](http://jonisalonen.com/2013/deriving-welfords-method-for-computing-variance/)

> Do not use this method to compute variance __ever__. This is one of those cases where a mathematically simple approach turns out to give wrong results for being numerically unstable. In simple cases the algorithm will seem to work fine, but eventually you will find a dataset that exposes the problem with the algorithm.

To put this in code. The naive approach vs. an online.combination after Welford. I used [StatReport](https://github.com/libmir/mir/pull/90/files) here, it's a OutputRange for mean + variance, but[Summary](https://github.com/DlangScience/dstats/blob/master/source/dstats/summary.d#L839) of dstats is nice as well.

```d
auto arr = [10e19 + 4, 7, 10e9 + 13, 10e15 + 16];
writefln("Naive: %f", arr.mean);
writefln("Online: %f", arr.stat.mean);
```

yields:

```
Naive: 25002500002500001792.000000
Online: 25002500002500000008.000000
```

For reference, it should be [25002500002500000010](http://www.wolframalpha.com/input/?i=mean+%7B10e19+%2B+4,+7,+10e9+%2B+13,+10e15+%2B+16%7D).
Or [if you prefer to read a paper](http://people.ds.cam.ac.uk/fanf2/hermes/doc/antiforgery/stats.pdf) (the simple trick is equation 4 in chapter 1):

![image](https://user-images.githubusercontent.com/4370550/27997973-b72b0e36-6503-11e7-8018-4b698c4006c9.png)

This method is more precise, but not as efficient due to division in the loop. However, I argue that most users won't care and correctness is more important than performance for a standard library.